instructions for dashboard staging (#336)
hubverse-org/hubDocs: docs/source/developer/dashboard-staging.md

and confirm the results are what you expected AND that nothing adverse happens
for dashboards that opt out of new features. **This is a practice known as
[integration testing](https://www.geeksforgeeks.org/software-engineering-integration-testing/).**

**This is not an exact science and you should use your best judgement when
moving changes into production.** The dashboards do have a lot of moving
pieces, but they are not infinite and not insurmountable. This chapter should
give you a few scenarios you can use to piece together the process for
confidently adding new features or fixing critical bugs in the dashboard
workflow.

### Generic steps of workflow tools

**An important concept to understand is that [the control
room](https://github.com/hubverse-org/hub-dashboard-control-room) workflows are
just that: workflows.** They implement the same basic steps as the [local
dashboard workflow](./dashboard-local.md) and push the individual outputs to
branches of the repository. In order to do that, they use the following steps:

1. install the tool that's required
2. download the dashboard configuration file
3. download the additional resources needed

> 4. run a the command to generate output

Reviewer: ```suggestion
4. run the command to generate output
```

5. push that output to a branch

> When thinking about staging the changes

Reviewer 2: Incomplete sentence?

Author: Incomplete sentence.

:::{admonition} Examples

For example in building the forecast data, the steps are:

1. install [hub-dashboard-predtimechart](https://github.com/hubverse-org/hub-dashboard-predtimechart)
2. download `predtimechart-config.yml` and `site-config.yml` from the dashboard repo
3. download the hub (defined in `site-config.yml`)
4. run `ptc_generate_target_json_files` and `ptc_generate_json_files`
5. save the output in the `ptc/data` branch and push it

similarly, for building the site, the steps are:

> 1. enter the [hub-dash-site-builder](https://github.com/hubverse-org/hub-dash-site-builder)

Reviewer: Would launch be a better word for this? In any case I think it would be useful to mention you're talking about a docker container.

Author: ```suggestion
1. launch an interactive shell from the [hub-dash-site-builder](https://github.com/hubverse-org/hub-dash-site-builder)
```

3. create a fork of a dashboard that would point to this new branch in the control room

After merging [the updated dockerfile with new tests and interface](https://github.com/hubverse-org/hub-dash-site-builder/pull/21) into the main branch of the repository, I created the docker image from the main branchby using the [Create, Test, and Publish Docker Image workflow](https://github.com/hubverse-org/hub-dash-site-builder/actions/workflows/build-container.yaml). Note that this docker image has the `main` tag[^image-name], but it is not published.

[^image-name]: This was before we had fully ironed out the details of the docker publishing workflow, so the image tag is actually `znk-dispatch-fix-34`. I am writing this as if we had the workflow that we have now.

Once I did that, I opened [hubverse-org/hub-dashboard-control-room#59](https://github.com/hubverse-org/hub-dashboard-control-room/pull/59) and then modified the workflow to use the new container and the new interface (see [hubverse-org/hub-dashboard-control-room@58628f0f4](https://github.com/hubverse-org/hub-dashboard-control-room/pull/59/commits/58628f0f4cef7e5c2cb5dc9455dc63235f47ec1d)).

> I think created a fork of the dashboard repositories and changed their `build-site.yaml` workflows to use the branch I was testing[^testing-branch]

Author: It was 'think' in the dining room with a candlestick. 

```suggestion
I created a fork of the dashboard repositories and changed their `build-site.yaml` workflows to use the branch I was testing[^testing-branch]
```

[^testing-branch]: In this process, I actually modified the workflow for the
    app, ran it without pushing, and inspected the artifacts. This had the same
    effect, but meant that I could test several repositories at once.

```diff
-uses: hubverse-org/hubverse-org/hub-dashboard-control-room/.github/workflows/generate-site.yaml@main
+uses: hubverse-org/hubverse-org/hub-dashboard-control-room/.github/workflows/generate-site.yaml@znk/use-release-hsdb/58
```

I then ran the workflows from the dashboard forks to confirm that the site was correctly generated.

:::

## Local Staging strategies

The following sections will cover local staging strategies based on what

> component you are modifying working. Note that you will often have to mix

Reviewer 2: type: s/b "based on what component you are modifying"?

Author: and where is the typo?

Reviewer 2: Sorry:
- "modifying working" didn't parse
- s/b = should be

Author: Ah I see! Thank you for the clarification. For future typos, using the "suggest edits" button is really helpful for me to know what you would like changed.

 
```suggestion
component you are modifying. Note that you will often have to mix
```

![screenshot of in-progress comment with suggest edits button (a page with a plus and minus symbol) highlighted](https://github.com/user-attachments/assets/ee8b5bd5-f5ab-4656-a467-5d39a01ba0fa)

[^image-name]: This was before we had fully ironed out the details of the docker publishing workflow, so the image tag is actually `znk-dispatch-fix-34`. I am writing this as if we had the workflow that we have now.

Once I did that, I opened [hubverse-org/hub-dashboard-control-room#59](https://github.com/hubverse-org/hub-dashboard-control-room/pull/59) and then modified the workflow to use the new container and the new interface (see [hubverse-org/hub-dashboard-control-room@58628f0f4](https://github.com/hubverse-org/hub-dashboard-control-room/pull/59/commits/58628f0f4cef7e5c2cb5dc9455dc63235f47ec1d)).

I created a fork of the dashboard repositories and changed their `build-site.yaml` workflows to use the branch I was testing[^testing-branch]

[^testing-branch]: In this process, I actually modified the workflow for the

>     app, ran it without pushing, and inspected the artifacts. This had the same

Author: It's definitely a strange balance to do this because 

1. I haven't yet written the page for the app 
2. people have _strong_ feelings about it
3. If I don't mention it, the example does not make sense

I'll see what I can do.

then the users downstream will have that version delivered via the CDN, but if
we turn around and release version 2.0.0, they will still get the 1.1.0 version.

When you release a JavaScript module, <https://cdn.jsDeliver.com> will pick it
up within 12 hours, but user machines can keep it cached for up to seven days.

If you want your changes to show up near instantly, you can [purge jsDeliver's CDN cache](https://www.jsdelivr.com/tools/purge) by entering two URLs.

The way to do this is to **paste the URLs in the browser and replace `cdn` with `purge`**.
Note that you also need to do this for **the un-versioned URL** as well:

```{code-block}
:caption: URLs to clear the cache for PredTimeChart: copy and paste into your browser
https://purge.jsdelivr.net/gh/reichlab/predtimechart@v3/dist/predtimechart.bundle.js
https://purge.jsdelivr.net/gh/reichlab/predtimechart/dist/predtimechart.bundle.js
```

```{code-block}
:caption: URLs to clear the cache for PredEvals: copy and paste into your browser
https://purge.jsdelivr.net/gh/hubverse-org/predevals@v1/dist/predevals.bundle.js
https://purge.jsdelivr.net/gh/hubverse-org/predevals/dist/predevals.bundle.js
```

:::

> ### Fetching orphan branches

Author: ```suggestion
### Accessing pre-built data
```

4. merge the control room branch to main and it will be live

(staging-tools-predevals)=
### hubPredEvalsData-docker

To stage changes to hubPredEvalsData-docker from the control room:

1. implement change in
   [hubPredEvalsData-docker](https://github.com/hubverse-org/hubPredEvalsData-docker)
   and push to the `main` branch after testing.
2. publish a new image from the main branch (note that only tags will create an official release, so this is safe to do)
2. create new branch in the control room and modify
   [`generate-data.yaml`](https://github.com/hubverse-org/hub-dashboard-control-room/tree/main/.github/workflows/generate-data.yaml) so that it points to the `main` branch instead of `latest`:
   ```diff
    runs-on: ubuntu-latest
    container:
   -  image: ghcr.io/hubverse-org/hubpredevalsdata-docker:latest
   +  image: ghcr.io/hubverse-org/hubpredevalsdata-docker:main
      ports:
        - 80
      volumes:
        - ${{ github.workspace }}:/project
   ```
3. fork a dashboard repository, point its workflows to the new control room
   branch
4. generate the data
5. generate the site and preview (NOTE: If the JavaScript component also

>    changes, you will need to preview the site locally)

Reviewer: Sorry for the silly question. Why do we also need to preview the site locally for changes in the JavaScript component? Won't remote testing achieve this? Or the point is we would need to also follow all the steps associated with testing changes to the JavaScript components?
